### ISSUE TITLE
Using Environments without creating deployment automatically

### ISSUE THREAD

**Issue by LaurenzReitsam** [actions/runner#2120]
**Current Situation:**
Every pipeline step that uses environments automatically creates a new deployment. [This seems to be wanted behavior](https://docs.github.com/en/actions/deployment/targeting-different-environments/using-environments-for-deployment#how-environments-relate-to-deployments).

**Problem**
Access to an ``environment`` might be also needed for other reasons than deployments. Like running integration tests (deployment already done; we want to assure correct behavior of latest deployment)

**Possible Solution**
Can we add an option to avoid an automatic deployment always when using ``environment``?
An idea might be to set an environment variable like `AUTO_DEPLOYMENT=false`.

**Additional information**
- [GitLab has option to specify what kind of action is done within environment](https://docs.gitlab.com/ee/ci/environments/#access-an-environment-for-preparation-or-verification-purposes)

**Comment by flobernd** [comment]
The exact same thing bugs me as well.

I would even go one step further and decouple the deployments from environments completely by default.

Imho every workflow should be able to be a deployment task - or not. Instead of using the environment to determine this, we could just have another string key in the yaml (similar to the "concurrency" key).

**Comment by civitaspo** [comment]
I have the same problem. I want to limit deployments only on protected branches after the pull request has been merged, but I want to use environment secrets at the time of the pull request as well. Specifically, it is a case of executing terraform plan in a pull request and terraform apply in an action that runs on a protected branch after the pull request is merged. Currently, it is not possible to limit the deployment to a protected branch while sharing the environment secrets to actions on non-protected branches.

**Comment by jameslounds** [comment]
Temporary Workaround: use github's api in actions to delete all deployments which match our sha.
We can use the [github-script](https://github.com/actions/github-script) action., which lets us use[ octokit.js](https://octokit.github.io/rest.js/v19) in the action 
Our job will need the `deployments: write` permission.

```
- name: Delete Previous deployments
        uses: actions/github-script@v6
        with:
          script: |
            const deployments = await github.rest.repos.listDeployments({
              owner: context.repo.owner,
              repo: context.repo.repo,
              sha: context.sha
            });
            await Promise.all(
              deployments.data.map(async (deployment) => {
                # we can only delete inactive deployments, so let's deactivate them first
                await github.rest.repos.createDeploymentStatus({ 
                  owner: context.repo.owner, 
                  repo: context.repo.repo, 
                  deployment_id: deployment.id, 
                  state: 'inactive' 
                });
                return github.rest.repos.deleteDeployment({
                  owner: context.repo.owner,
                  repo: context.repo.repo,
                  deployment_id: deployment.id
                });
              })
            );
```

**Comment by tianhuil** [comment]
Same: would love to be able to disable with a yaml property like `- autodeploy: false`.

**Comment by berkeli** [comment]
> Temporary Workaround: use github's api in actions to delete all deployments which match our sha. We can use the [github-script](https://github.com/actions/github-script) action., which lets us use[ octokit.js](https://octokit.github.io/rest.js/v19) in the action Our job will need the `deployments: write` permission.
> 
> ```
> - name: Delete Previous deployments
>         uses: actions/github-script@v6
>         with:
>           script: |
>             const deployments = await github.rest.repos.listDeployments({
>               owner: context.repo.owner,
>               repo: context.repo.repo,
>               sha: context.sha
>             });
>             await Promise.all(
>               deployments.data.map(async (deployment) => {
>                 # we can only delete inactive deployments, so let's deactivate them first
>                 await github.rest.repos.createDeploymentStatus({ 
>                   owner: context.repo.owner, 
>                   repo: context.repo.repo, 
>                   deployment_id: deployment.id, 
>                   state: 'inactive' 
>                 });
>                 return github.rest.repos.deleteDeployment({
>                   owner: context.repo.owner,
>                   repo: context.repo.repo,
>                   deployment_id: deployment.id
>                 });
>               })
>             );
> ```
Works nicely, but the following line needs to be removed for it to work: 

```
# we can only delete inactive deployments, so let's deactivate them first
```

**Comment by yusijs** [comment]
This would be great, my pull requests currently look like this:
![image](https://user-images.githubusercontent.com/11261560/214039525-302c9a73-6e09-4d7d-8a8d-cf0714bcb792.png)


We use environments for builds etc as well (secrets), so it becomes a mess very quickly. Being able to specify the environment at the top level (before the jobs) might also help a bit, but ideally it would be possible to do something like this in the job definition:
```
environment:
  name: dev
  url: https://github.com
  deployment: false
```

**Comment by SrBrahma** [comment]
@jameslounds what about creating an Action for it in the Marketplace?

**Comment by constgen** [comment]
We already have a ton of actions for deployment creation and status update. It is not a necessary to create a new one. But my issue now is when I use a custom deployment actions for more control of deployment status, I have a duplicated deployment in history. Having one of these would help:
- A variable in Github Context with `deployment_id`, so actions can check that `environment` was used in configuration and reuse existing `deployment_id` instead of creating a new one.
- Option to disable automatic deployment report when `environment` is defined in a job

**Comment by nagibyro** [comment]
Just FYI if you're using @jameslounds workaround for workflows that run for pull requests, you need to modify the script, since the `GITHUB_SHA` is the last merge commit not the head commit that just got pushed to the PR branch (which is what the deployments are created with). You can add the `github.event.pull_request.head.sha` as an environment variable to the action:

```yaml
  delete_github_deployments:
    runs-on: ubuntu-latest
    needs: run_tests
    if: ${{ always() }}
    steps:
      - name: Delete Previous deployments
        uses: actions/github-script@v6
        env:
          GITHUB_SHA_HEAD: ${{ github.event.pull_request.head.sha }}
        with:
          script: |
            const { GITHUB_SHA_HEAD } = process.env
            const deployments = await github.rest.repos.listDeployments({
              owner: context.repo.owner,
              repo: context.repo.repo,
              sha: GITHUB_SHA_HEAD
            });
            await Promise.all(
              deployments.data.map(async (deployment) => {
                await github.rest.repos.createDeploymentStatus({ 
                  owner: context.repo.owner, 
                  repo: context.repo.repo, 
                  deployment_id: deployment.id, 
                  state: 'inactive' 
                });
                return github.rest.repos.deleteDeployment({
                  owner: context.repo.owner,
                  repo: context.repo.repo,
                  deployment_id: deployment.id
                });
              })
            );
```

We had to add this as a job that ran at the end of the workflow instead of the last step of other jobs because the deployment didn't seem to always be deleted in that case. So you still see the message on the PR's till the whole pipeline runs which can still be confusing for folks but seems like the best you can do for now.

**Comment by SrBrahma** [comment]
@nagibyro So if I have 4 .yaml files, I add this to their end?

**Comment by kylebjordahl** [comment]
I ran into a similar version of this where our deploy workflows are manually triggered (either via UI or GH CLI) and we supply a specific release ref to deploy; the deployments created by the workflow show the current state of main as the deployed ref, when in fact they should show the ref of the specified release.

I made [this Action](https://github.com/kylebjordahl/replace-workflow-deployment) to handle our case, but feel free to fork it for your own purposes; hope it helps others until this gets sorted

**Comment by amos-kibet** [comment]
This action worked for me. It optionally deletes deployments and environments. You decide what to delete.
[strumwolf/delete-deployment-environment](https://github.com/strumwolf/delete-deployment-environment)

**Comment by kamilzzz** [comment]
Any comments from the GitHub side regarding this issue? It has been reported 9 months ago.

Workarounds presented here work but they are well... workaround and I feel like this should be implemented by the GitHub Actions itself. There are several scenarios where workflow may target environment but is not a real deployment, like running Terraform plan or running integration tests as mentioned before.

**Comment by Hronom** [comment]
Same here, any answer for Enterprise customer?

**Comment by fabasoad** [comment]
Have the same problem. Also, on GHES. Would be great to have any updates on this. Thanks.

**Comment by MichaelMHoff** [comment]
Same problem here. We are running automated selenium tests with stage-dependent credentials and these "fake deployments" really mess things up as they are not only displayed for PRs, but also in our Jira integration

**Comment by danielsantiago** [comment]
I'm having the same problem with the Jira integration, each Job in the Workflow is treated as a different deployment.

**Comment by craig-king** [comment]
+1 from here as well. Terraform plans are a good example of targeting an environment but not "deploying" to it. We really need a flag to specify if this is a full deployment or not

**Comment by JonathanAtCenterEdge** [comment]
+1, our workflows to bring down our infrastructure act as deployments due to this.

**Comment by bombillazo** [comment]
Our account is bloated with ephemeral environments/deployment simply because we want to use some environment variables as a setup. 

It would be great to auto-delete them after the PR is merged.

**Comment by hknutsen** [comment]
> +1 from here as well. Terraform plans are a good example of targeting an environment but not "deploying" to it. We really need a flag to specify if this is a full deployment or not

This is my exact issue as well. I run "Terraform Plan" for multiple environments on PRs, and since they all count as deployments, it pollutes the PR.

**Comment by tinogo** [comment]
> > +1 from here as well. Terraform plans are a good example of targeting an environment but not "deploying" to it. We really need a flag to specify if this is a full deployment or not
> 
> This is my exact issue as well. I run "Terraform Plan" for multiple environments on PRs, and since they all count as deployments, it pollutes the PR.

Same here!
Things get even more annoying, when you've configured deployment approvals, too. In these situations you also have to approve each and every job which uses the environment, but isn't really deploying anything...

Therefore: +1 for the `deployment: false`-property or something like this.

**Comment by AustinZhu** [comment]
> +1 from here as well. Terraform plans are a good example of targeting an environment but not "deploying" to it. We really need a flag to specify if this is a full deployment or not

Exactly!

**Comment by mkarbo** [comment]
https://github.com/orgs/community/discussions/36919#discussioncomment-6852220 see also this

**Comment by TrevorSmith-msr** [comment]
I have this same issue with a workflow that runs Terraform Plan.

**Comment by bombillazo** [comment]
We used the script suggested here (using the GH API) to manually delete the environments periodically. It's really bad this isn't available as an option built into the GH repo...

**Comment by godd9170** [comment]
We are getting great value out of the environment specific environment variables for our frontend tests/builds, and are also continuously being confused about the deployment statuses. Is there a way perhaps to indicate the context for the job, so we can access the right vars, without using the `environment` key in it's definition?

**Comment by mikocot** [comment]
I feel like those 2 issues are related:
https://github.com/orgs/community/discussions/67727

https://github.com/orgs/community/discussions/67728

**Comment by johanmolen** [comment]
Would be great to have any updates on this.

**Comment by tornike** [comment]
In my case deployments are happening from separate IaC repository and in application repos environments are used just for the context, _github deployments_ there has no meaning other than to cause inconvenience and misunderstanding. There should be some option to disable automatic _github deployment_ creation.

**Comment by piotrekkr** [comment]
This issue makes using environment scoped vars and secrets pretty irritating process that pollutes repository with useless deployments that could never actually happen. 

Here is my use case:

- I have three environments like `staging`, `sandbox`, `production`.
- I have deploy workflows for each environment that triggers multiple jobs that use environment scoped vars/secrets (e.g `initialize`, `build`, `deploy`, `finalize` jobs).
- I have some workflows that do not deploy anything but also need to use env scoped vars/secrets (e.g. to connect to cloudsql instance and run some maintenance tasks on selected environment).
- I create and update deployments manually when needed because I need to set custom payload on creation and do this when some conditions are met (this is not possible using environments).

As I run deploy workflow I get three deployments. First is manual one created by me, next from `build` job and one from `deploy` job. Since my manual deployment was older than the `deploy` job created it is not shown as latest deployment on env. I needed to create custom action to remove those unwanted deployments.

Same thing with other workflows that do not deploy anything but wants to access environment. Each run generates new deployment and I need to cleanup. 

Why pollute such useful feature like environment scoped vars and secrets with deployments?

**Comment by airtonix** [comment]
we have many apps in our monorepo.

each app gets a set of environments like: 

- `dev/{hosting_platform}/{appname}`
- `staging/{hosting_platform}/{appname}`
- `prod/{hosting_platform}/{appname}`

each contains a set of variables that control things from aws account id, build command all the way through to slack channel id.

During a actual deployment we like what this does: 

```yml

environment: `${{inputs.Stage}}/s3/${{inputs.AppName}}`

steps:
  - uses: some/action
    with: 
      SlackChannelId: ${{vars.SLACK_CHANNEL_ID}}

...
```

but we have other workflows that are either on PRs or manually run workflows that are sensitive to the AppName.

Here, like everyone else in this thread; we find the association with the concept of a deployment to be counterproductive for this major reason: 

Jira Integration with Github treats a "Github Deployment" as a literal Deployment. so they end up on the Releases board and are marked as a release on tickets associated with the commits involved.

This ends up causing a lot of wasted time explaining things over and over again.

There is a need to be able to define access to variables through the `environment` key without it triggering a deployment association.

Perhaps this semantic association should be removed as an implicit behaviour; Instead this association should be created by running an explicit action.

**Comment by Theodlz** [comment]
I just started using secrets for github actions and honestly... I don't quite get why this is automatically coupled to deployments. I see how it can be useful for some, but most just want to use secrets for their CI, and the PRs flooded with "deployment inactive" messages make it somewhat painful to use. I mean, there is nothing being deployed whatsoever, and its confusing for the developer and the reviewer of a PR.

**Comment by jelmerk** [comment]
+1 for this. It makes no sense to tie these two things together

**Comment by AndresPinerosZen** [comment]
In my use case I have no problem if the deployment is created, but I would like to have control over the status of the deployment once the job is finished. I don't want the status to be "green", I'd like to leave it as "in-progress" so other tools can work with the status.

**Comment by aheruz** [comment]
+1

**Comment by kumarpramod** [comment]
+1

**Comment by hikerspath** [comment]
Functionally, `environments` should not equal a `deployment`. These trigger different resultant code to tools like MS DevOps, Jira, and others that track deployment events. It is functionally possible to require environment-based secrets but perform something like a terraform plan or other test suite which is not performing any action itself. I agree that there should be an override option of `deployment: false` or some sort of thing.

**Comment by joelcoxokc** [comment]
Hello... Any updates on this.. We spent a lot of time moving everything over to environments using git actions.. And now we have this massive mess with 15 deployment events being created from one deployment... 

We need to environment secrets in each of 15 different actions... 

But now we have 15 staging/production deployments... when it should only be one.. 

This should only be triggered after our final run. Please Please help!

**Comment by Anemony22** [comment]
We're in the same boat. Would be great to be able to use the power of environment secrets without spamming the PR history. 

We currently use environment secrets in actions to do a bunch of pre-deployment checks.

**Comment by nicolasguillen-tamedia** [comment]
+1

**Comment by kylebjordahl** [comment]
FWIW, we're still having this issue in an entirely new project (I commented on this over a year ago while working on a totally different product) so I solved it in a different way: **use the `gh` cli in a composite action to fetch the environment's vars from the API.**

It's not terribly elegant, and I had to set up an internal GitHub App in order to get a token with the proper permissions, but now that I have it set up it works like a charm. While I could make a reusable action to do this, most of the work is actually in the process of setting up the GitHub App and dealing with the token, which you'd have to do yourself anyway.

Would be great if GitHub just allowed us to specify whether a workflow should create a deployment or not 😢

**Comment by andreujuanc** [comment]
I'm thinking on ditching the environments and go back to using my fork of [this](https://github.com/noliran/branch-based-secrets)

**Comment by airtonix** [comment]
> I'm thinking on ditching the environments and go back to using my fork of [this](https://github.com/noliran/branch-based-secrets)

Environments are more than just secrets. They also provide the manual approval step

**Comment by andreujuanc** [comment]
> Environments are more than just secrets. They also provide the manual approval step
I know, but we can't even use it for PRs. It's almost like GH devs don't use GH products.

**Comment by ricardofalc** [comment]
This is very, very annoying.
Is it possible to introduce a  yaml property to disable this?

![image](https://github.com/user-attachments/assets/395b7d4a-d969-4da8-9eed-188b99d734f7)
![image](https://github.com/user-attachments/assets/87afcc82-037c-4f21-b712-65e0fc912bd8)

**Comment by GrantBirki** [comment]
👋 Hey folks! I work here at GitHub. I'm on the security side of things here and do not work directly on Actions but I do maintain quite a few open source GitHub Actions. Having a feature like this one would be especially helpful for an Action that I maintain ([github/branch-deploy](https://github.com/github/branch-deploy)).

I'm going to start down the path of contributing to a few internal services here at GitHub to see if I can make this feature a reality for us.

My end goal is a non-breaking change that would allow for something like this:

```yaml
environment:
  name: dev
  url: https://github.com
  deployment: true # <-- this would be the new parameter
```

By default, the new `deployment:` parameter would be set to `true` so that this could be a non-breaking change. Then we can set it to `false` in our workflows where we do not want environments to deploy automatically.

Wish me luck!

> I will update back here with any progress or further info as I'm able to

**Comment by qoomon** [comment]
@GrantBirki an additional global workflow config variable would be awesome to prevent explicitly disable deployment for every environment usage. I have no workflow where an automated deployment make sense. I rather call an action to explicitly create and update a deployment, so I can decide when a deployment has taken place.

**Comment by qoomon** [comment]
Or even introduce a workflow version field to cover breaking changes.

**Comment by chemipaine** [comment]
Hey @GrantBirki any news on your feature to solve the issue? Actually, for us is a big issue as well as we run test job after deployment and is needed to use the same secrets as the deploy, in other hand we have manual approval so basically, we need to approve two times and have two deployments instead one, which is not the nicest behavior tbh...

**Comment by GrantBirki** [comment]
@chemipaine Just to be 100% transparent I will be a bit slow to work on this issue. The repos/projects that I am interacting with I have not interacted with before. That being said, I have made a bit of progress internally. I still need to finish communicating with teams to ensure the changes I have proposed are correct and to get all the proper sign-offs. So far so good, but it is going to take some time 🙇

**Comment by chemipaine** [comment]
Thank you so much for coming back on this @GrantBirki, in the meantime i produced a slightly twisted solution for the approvals issue, which for us was the biggest problem. We created two dummy environments, one for prod approvals and other for non-prod, and made an action to use in any part of the pipelines to stop them for approval. This is an example on how it would work:

CHILD PIPELINE:

```shell
name: "Pipeline Approvals Step"
on:
  workflow_call:
    inputs:
      environment:
        required: true
        type: string

jobs:
  approvals_step:
    name: Approval Request
    runs-on: ubuntu-latest
    environment:
      name: ${{ inputs.environment }}
    steps:
      - name: Convert Environment Name
        id: convert_environment
        run: |
          if [ "${{ inputs.environment }}" = "approvals-prd" ]; then
            echo "APPROVAL_MESSAGE=***IMPORTANT*** You have APPROVED changes to the PRODUCTION environment! Please ensure all modifications are implemented with EXTREME CARE!***" >> $GITHUB_ENV
          elif [ "${{ inputs.environment }}" = "approvals-nonprd" ]; then
            echo "APPROVAL_MESSAGE=You have approved the changes to the non-production environment. Feel free to proceed with the updates." >> $GITHUB_ENV
          else
            echo "APPROVAL_MESSAGE=Changes will be made to the specified environment." >> $GITHUB_ENV
          fi

      - name: Send Approval Request
        id: approval_request_message
        run: |
          echo "$APPROVAL_MESSAGE"

``` 

PARENT PIPELINE TO CALL CHILD:

```shell
name: "Trigger Approval Request"

on:
  workflow_dispatch:

jobs:
  trigger_approval:
    name: Approval Requirement
    uses: my-repository/.github/workflows/approvals.yml@main
    with:
      environment: "approvals-prd" #You need to change this to point to prod or non-prod or your usecase
``` 

Hopes this help someone with same issue, not the nicest solution but at least you don't must approve in the same pipeline several times when you need to run more than one job using the environment variables...

**Comment by JonasSonn** [comment]
Thanks @GrantBirki! You're making a Christmas miracle come true ❤️

**Comment by Steve-Glass** [comment]
👋 I'm with the Actions product team and have been working with @GrantBirki on getting this prioritized. We've just started planning for the Jan-Mar timeframe and have bubbled this up for prioritization. No Christmas miracle 😢 However I will report back mid January with the outcome.

**Comment by Chippiewill** [comment]
> My end goal is a non-breaking change that would allow for something like this:
> 
> environment:
>   name: dev
>   url: https://github.com
>   deployment: true # <-- this would be the new parameter
> By default, the new `deployment:` parameter would be set to `true` so that this could be a non-breaking change. Then we can set it to `false` in our workflows where we do not want environments to deploy automatically.

You may wish to consider not having it just be a simple boolean. Gitlab's approach to this is to have an 'action' field which accepts a limited set of string values:

- start (creates a deployment)
- prepare (just access variables/secrets)
- stop (stops a deployment)
- verify (just access variables/secrets)
- access (just access variables/secrets)

This captures additional semantic information which Gitlab have subsequently exposed into their UI and makes the schema more flexible to changes down the line. You wouldn't need to expose as many options as Gitlab has done initially (Gitlab originally only had start and stop themselves), just enough to cover this initial usecase (e.g. start and prepare, or maybe some different naming).

**Comment by jmaciejewski-Roadpost** [comment]
> 👋 I'm with the Actions product team and have been working with [@GrantBirki](https://github.com/GrantBirki) on getting this prioritized. We've just started planning for the Jan-Mar timeframe and have bubbled this up for prioritization. No Christmas miracle 😢 However I will report back mid January with the outcome.

Hi, are there any new updates?

**Comment by Steve-Glass** [comment]
> Hi [@Steve-Glass](https://github.com/Steve-Glass), are there any new updates?

This item is on our team's backlog for the April-June 2025 timeframe. While we attempted to prioritize it after the holidays, priorities shifted. I will provide a more refined estimate for completion after planning officially concludes for our team in March.

Thank you for your patience!

**Comment by duaneking** [comment]
Any updates?

Due to how broken these features are, there are large parts of GitHub we simply cant trust enough to use. Please prioritize these at a higher level.

**Comment by wmattei** [comment]
Hi @Steve-Glass. I am happy to hear that this is in the team backlog. 
Is this part of the [GH public roadmap](https://github.com/orgs/github/projects/4247), or is it kept somewhere internally?

Thanks!

**Comment by nebuk89** [comment]
@wmattei at the moment the public roadmap is being kept for 'bigger releases' only rather than smaller items. As a result we are working on finding a home for items like this to communicate better :( I am sorry we don't have something currently.

Right now 🤞 we are still on track for 'no later than the end of June', thanks for all continuing to engage here.

**Comment by tw-sematell** [comment]
Since I have not seen a documentation update yet, I guess there's no update yet?

**Comment by afzanjamalreach** [comment]
> Right now 🤞 we are still on track for 'no later than the end of June', thanks for all continuing to engage here.

@nebuk89 any updates you can share?

**Comment by tdupoiron** [comment]
TotalEnergies is waiting for this great enhancement. Hope it can be shipped soon 🚀
cc @stephanebiermann

**Comment by svivekkrishna** [comment]
I'd really love this feature too. As noted elsewhere, we use the environment to scope workflows and variables for execution of automation that isn't really a deployment.

Is this coming out any time soon?

**Comment by christiaan-bronkhorst** [comment]
> [@wmattei](https://github.com/wmattei) at the moment the public roadmap is being kept for 'bigger releases' only rather than smaller items. As a result we are working on finding a home for items like this to communicate better :( I am sorry we don't have something currently.
> 
> Right now 🤞 we are still on track for 'no later than the end of June', thanks for all continuing to engage here.

@nebuk89 It is July now, can we please just get some update?

**Comment by whatasoda** [comment]
@nebuk89 Is there any update? 👀

**Comment by ocbaker** [comment]
~~Valve~~ GitHub time works in mysterious ways. It's be great to get an update for this feature!

Our teams issue is we build production images as part of any merge into main, but we deploy them in a separate workflow so the deployments have nothing to do with actually deploying lol.

**Comment by sekhavati** [comment]
Please can you confirm if the upcoming changes will also incorporate what is discussed [here](https://github.com/orgs/community/discussions/58020)? 

ie: not needing multiple approvals across different jobs that use the same `environment`, when say a specific job is only using the environment to gain access to its secrets/variables and not a "deployment" as such

**Comment by GrantBirki** [comment]
Hey 👋 its me again! You might remember when I tried to implement this feature [almost a year ago](https://github.com/actions/runner/issues/2120#issuecomment-2432890056). I tried my very best but my window of opportunity passed where I had free cycles to complete this work (I'm on security and don't work directly on Actions but I want this feature too).

I had a draft PR that implemented the base components of this feature but it didn't do everything _quite right_. I'm going to ping some folks internally to see where things are at with this, and if maybe I can get some extra eyes on my (now very old) PR. I do know that some things shipped recently that are somewhat related to this change that might help support efforts around this again.

> Please note: I am not a PM, I don't directly work on Actions, and I may not have the answer you all are looking for but I care about this product and this is a papercut that I want to see go away for the rest of the community so I'll do my best and see if I can get movement here again! 🙇

**Comment by ayodejiayodele** [comment]
Thanks team. We are still looking forward to this feature. My customers are keen!

**Comment by kruncher** [comment]
This is an issue for us as well; we need to have env specific secrets to push to a different container registry based on the env.

Since we have a matrix job it ends up flooding our PRs with these false deployment notifications which makes PRs very noisy.

As mentioned by someone else above; our case is a mono repository as well.

It would be good if we could just set `isDeployment: false` when specifying `environment`.

Or alternatively, if there was a way for us to read the env secrets one time in a prepare step and then securely share that secret with various other jobs that occur within our workflow. But there doesn't seem to be a safe way to share those secrets. The deployment notice would still be incorrect; but at least it would only appear once instead of many times per synchronize on a PR.

**Comment by airtonix** [comment]
So there's three things the `environment: <<KEY>>` key does that I'm aware of: 

- lets you dynamically populate `${{ vars. }}` based on what you have in `Settings > Environments > KEY`
- Records a deployment as in progress, etc blah blah. (the thing that's annoying everyone here)
- Gives you the ability to Require approval if a job attempts to run for a particular `environment: KEY`

If you're just doing 1️⃣ , then you can use the following api call: 

https://gist.github.com/airtonix/b50e65b83624436a06fce92d3f208541

With this you can access particular values in your Environment secrets or variables without associating the run with a deployment.

Run it in your workflows, create an action, what ever.

**Comment by gplessis** [comment]
Also, if not all the runs of your workflow require to set it, you can unset the `environment` property to avoid creating unnecessary deployments:

```
    environment: ${{ startsWith(github.event.ref, 'refs/tags/') && 'staging' || null }}
```

This way, we managed to reduce the noise on our PRs.

**Comment by andreujuanc** [comment]
> Also, if not all the runs of your workflow require to set it, you can unset the `environment` property to avoid creating unnecessary deployments:
> 
> ```
>     environment: ${{ startsWith(github.event.ref, 'refs/tags/') && 'staging' || null }}
> ```
> 
> This way, we managed to reduce the noise on our PRs.

Yes but then you wont be able to get secrets from the environment. Right?

**Comment by gplessis** [comment]
> Yes but then you wont be able to get secrets from the environment. Right?

Correct. Our use case (and maybe others') for setting the environment is to have [our GitHub workflow and its identity provider to be trusted by its mapped managed identity in Azure](https://learn.microsoft.com/en-us/entra/workload-id/workload-identity-federation-create-trust-user-assigned-managed-identity?pivots=identity-wif-mi-methods-azp#environment-example).

**Comment by airtonix** [comment]
> Yes but then you wont be able to get secrets from the environment. Right?

you can just use this instead if that's all you wanted

https://gist.github.com/airtonix/b50e65b83624436a06fce92d3f208541

**Comment by piotrekkr** [comment]
> > Yes but then you wont be able to get secrets from the environment. Right?
> 
> you can just use this instead if that's all you wanted
> 
> https://gist.github.com/airtonix/b50e65b83624436a06fce92d3f208541

Using API does not help with env secrets. Those cannot be fetched from it afaik.

**Comment by minhdient49** [comment]
Github.com/truongminhdien833@gmail.com /https://www.youtube.com/@truongminhdien2-i6ghttps://youtu.com/truongminhdien833@gmail.com /Trương Minh Điền Telegram +84 0916363743 minh Dien truong (#### 9999999999999%%##/agribank  7501205196575 trương Minh Điền /

quedien47@gmail.com 

Truongminhdien833@gmail.com

**Comment by minhdient49** [comment]
Trương Minh Điền https://youtu.com/truongminhdien833@gmail.com /Trương Minh Điền Telegram +84 0916363743 minh Dien truong (#### 9999999999999%%##/agribank  7501205196575 trương Minh Điền /

quedien47@gmail.com 

Truongminhdien833@gmail.com  
>     environment: ${{ startsWith(github.event./truongminhdien833@gmail.com /') && 'staging' || null }}
> ```
> 
> This way, we managed to reduce the noise on our PRs.

**Comment by albertodeago** [comment]
@airtonix  can you share the `permissions` you set for that script to work?
I tried a lot of them but I always get a 403 (`gh: Resource not accessible by integration (HTTP 403)`)

How do you give the workflow (the `${{ github.token }}`) the permissions to make that request?

**Comment by nebuk89** [comment]
Sorry I haven't updated, this didn't ship (clearly), it's back on our cards going into the start of the new calendar year. Sorry we missed expectations here, we will do better.

**Comment by ayodejiayodele** [comment]
Thanks @nebuk89 , is there a new timeline we should look forward to?

**Comment by reemasardana** [comment]
👋 from Actions Engineering. We're actively working on this and targeting shipping by end of March.

**Comment by jreinstra** [comment]
![image](https://github.com/user-attachments/assets/45946c4e-a7fd-4aed-be89-8af22f4cce54)

**Comment by salmanmkc** [comment]
Thanks for waiting on this, this is now available, a changelog will be coming out as well, but here's a heads up.

Use `deployment: false` 

```yaml
jobs:
  test:
    runs-on: ubuntu-latest
    environment:
      name: staging
      deployment: false
    steps:
      - run: echo "Staging secrets accessible, no deployment created"
        env:
          API_KEY: ${{ secrets.API_KEY }}
```

**What's not compatible:** custom deployment protection rule apps (job will fail — remove `deployment: false` or the custom rule).

**Docs:**
- [Using environments without deployments](https://docs.github.com/en/actions/how-tos/deploy/configure-and-manage-deployments/control-deployments#using-environments-without-deployments)
- [Deploying to an environment](https://docs.github.com/en/actions/how-tos/write-workflows/choose-what-workflows-do/deploy-to-environment)
- [Custom deployment protection rules](https://docs.github.com/en/actions/how-tos/deploy/configure-and-manage-deployments/create-custom-protection-rules#approving-or-rejecting-deployments)
- [Changelog](https://github.blog/changelog/2026-03-19-github-actions-late-march-2026-updates/#github-actions-now-allows-developers-to-use-environments-without-auto-deployment)

**Comment by dbrezhniev** [comment]
Thanks for the update! Just curious, why apply protection rules (approvers, timers) if such a job is no longer a deployment?

**Comment by salmanmkc** [comment]
> Thanks for the update! Just curious, why apply protection rules (approvers, timers) if such a job is no longer a deployment?

Customers may still want approvals for accessing secrets, etc. Custom deployment protection rules rely on the payload, and since we won't be creating a deployment there would be no deployments object in the payload to send which can break existing custom apps.

There's the option to keep this current behaviour, or if we don't apply approvers or timers, it's a security risk or could be breaking, if we don't apply those rules, all existing workflow runs that have these gates will automatically be approved, let's say a workflow run requires someone to approve to access a secret, that's dangerous if we don't apply the rule suddenly, and someone can break it with just adding `deployment: false`. The alternative is to keep it as it is, or deny everything like I've done in the current implementation with custom apps (as per the reason above), however keeping it how it is allows the flexibility to be left to the customer.

**Comment by salmanmkc** [comment]
Changelog has been published: https://github.blog/changelog/2026-03-19-github-actions-late-march-2026-updates/#github-actions-now-allows-developers-to-use-environments-without-auto-deployment

**Comment by qoomon** [comment]
<img height="250px" src="https://media1.giphy.com/media/v1.Y2lkPWZjZGU1NDk1NHVqMTUyN2RsczVyNWdyaDlsOG9nNmNmYXE3YjhyZTl1d2k3d3JtdCZlcD12MV9naWZzX3NlYXJjaCZjdD1n/5VKbvrjxpVJCM/giphy.gif" width="312px" itemtype="http://schema.skype.com/Giphy" key="gif_0">

**Comment by GrantBirki** [comment]
@salmanmkc this is huge, thank you!! Did you end up landing the PR that I had started working on many many moons ago?

**Comment by airtonix** [comment]
> > > Yes but then you wont be able to get secrets from the environment. Right?
> > 
> > 
> > you can just use this instead if that's all you wanted
> > https://gist.github.com/airtonix/b50e65b83624436a06fce92d3f208541
> 
> Using API does not help with env secrets. Those cannot be fetched from it afaik.

It does. We use it now.

**Comment by salmanmkc** [comment]
> [@salmanmkc](https://github.com/salmanmkc) this is huge, thank you!! Did you end up landing the PR that I had started working on many many moons ago?

My pleasure, regards to your PR, didn't see that, just started this from scratch!

**Comment by salmanmkc** [comment]
Closing this issue, any more feedback, feel free to open a new issue!

**Comment by ianwinsemius** [comment]
@GrantBirki does this mean we don't have to 👍 every branch deploy now?! That would be amazing if true.

**Comment by GrantBirki** [comment]
@ianwinsemius that is actually a _different_ issue related to the `pushed_at` field being bound to commits in the API. @iolsen is the one I was chatting with about that feature.

By this issue getting closed, it doesn't change the 👍 on branch-deploy comments, but it does allow you to drop the need for using `environment: production-secrets` and instead you can just use:

```yaml
environment:
    name: production
    deployment: false # the branch-deploy action is responsible for creating the deployment - see: https://docs.github.com/en/actions/how-tos/deploy/configure-and-manage-deployments/control-deployments#using-environments-without-deployments
```

Since the branch-deploy Action is responsible for creating the deployment object in the API, we used a "hack" to keep the security protections of environments but by using two envs: `production` (for the main `deploy` workflows triggered on merge) and `production-secrets` that the `branch-deploy` workflows used so they didn't clobber the deployment object that the branch-deploy action creates.
